copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/WorkerCoordinator.java

```diff
@@ -0,0 +1,274 @@
+ */
+public final class WorkerCoordinator extends AbstractCoordinator implements Closeable {
+    private static final Logger log = LoggerFactory.getLogger(WorkerCoordinator.class);
+
+    // Currently Copycat doesn't support multiple task assignment strategies, so we currently just fill in a default value
+    public static final String DEFAULT_SUBPROTOCOL = "default";
+
+    private final KafkaConfigStorage configStorage;
+    private CopycatProtocol.Assignment assignmentSnapshot;
+    private final CopycatWorkerCoordinatorMetrics sensors;
+    private ClusterConfigState configSnapshot;
+    private final WorkerRebalanceListener listener;
+
+    private boolean rejoinRequested;
+
+    /**
+     * Initialize the coordination manager.
+     */
+    public WorkerCoordinator(ConsumerNetworkClient client,
+                             String groupId,
+                             int sessionTimeoutMs,
+                             int heartbeatIntervalMs,
+                             Metrics metrics,
+                             String metricGrpPrefix,
+                             Map<String, String> metricTags,
+                             Time time,
+                             long requestTimeoutMs,
+                             long retryBackoffMs,
+                             KafkaConfigStorage configStorage,
+                             WorkerRebalanceListener listener) {
+        super(client,
+                groupId,
+                sessionTimeoutMs,
+                heartbeatIntervalMs,
+                metrics,
+                metricGrpPrefix,
+                metricTags,
+                time,
+                requestTimeoutMs,
+                retryBackoffMs);
+        this.configStorage = configStorage;
+        this.assignmentSnapshot = null;
+        this.sensors = new CopycatWorkerCoordinatorMetrics(metrics, metricGrpPrefix, metricTags);
+        this.listener = listener;
+        this.rejoinRequested = false;
+    }
+
+    public void requestRejoin() {
+        rejoinRequested = true;
+    }
+
+    @Override
+    public String protocolType() {
+        return "copycat";
+    }
+
+    @Override
+    public LinkedHashMap<String, ByteBuffer> metadata() {
+        LinkedHashMap<String, ByteBuffer> metadata = new LinkedHashMap<>();
+        configSnapshot = configStorage.snapshot();
+        CopycatProtocol.ConfigState configState = new CopycatProtocol.ConfigState(configSnapshot.offset());
+        metadata.put(DEFAULT_SUBPROTOCOL, CopycatProtocol.serializeMetadata(configState));
+        return metadata;
+    }
+
+    @Override
+    protected void onJoin(int generation, String memberId, String protocol, ByteBuffer memberAssignment) {
+        assignmentSnapshot = CopycatProtocol.deserializeAssignment(memberAssignment);
+        if (assignmentSnapshot.failed()) {
+            log.warn("Join group completed, but assignment failed. Reading to end of config and retrying.");
```

Comment:
Should we have a way to backoff before trying to rejoin in this case? Seems like we would most likely still have the same config snapshot. Could the leader give us information about which config version to wait for?